### ISSUE TITLE
LaTeX rendering fails in jupyter notebook on master branch.

### ISSUE THREAD

**Issue by bjodah** [sympy/sympy#9799]
See:

http://nbviewer.ipython.org/urls/gist.githubusercontent.com/bjodah/b0a8a3be8eb566f8dfc7/raw/1b148fe374497a6fd00af9ff05d73acb77f7f228/bug.ipynb

In case above link goes 404:

In a Python 2.7 notebook using sympy master branch.
```
import sympy
from sympy.interactive import printing
printing.init_printing()
x = sympy.Symbol('x')
pw = sympy.Piecewise((1, sympy.Eq(x, 0)), (0, True))
pw
```

End of traceback:
```
/usr/local/lib/python2.7/dist-packages/matplotlib/mathtext.pyc in parse(self, s, fonts_object, fontsize, dpi)
   2343                         err.line,
   2344                         " " * (err.column - 1) + "^",
-> 2345                         str(err)]))
   2346         self._state_stack = None
   2347         self._em_width_cache = {}

ValueError: 
\begin{cases} 1 & \text{for}\: x = 0 \\0 & \text{otherwise} \end{cases}
^
Unknown symbol: \begin (at char 0), (line:1, col:1)
```

Possibly related to #9692

**Comment by bjodah** [comment]
#9801 did not resolve this particular issue.

**Comment by jksuom** [comment]
This is what happens.
```
\documentclass[10pt]{article}
\pagestyle{empty}
\usepackage{amsmath,amsfonts}\begin{document}
\begin{equation*}\begin{cases} 1 & \text{for}\: x = 0 \^@ & \text{otherwise} \end{cases}\end{equation*}

\end{document}
```

So ``\0`` is interpreted as a hex null character. This does not necessarily occur in SymPy, since the latex text comes via ipython/notebook.

**Comment by bjodah** [comment]
No I don't think what's happening here, I think matplotlib lacks support for \begin{cases} .... \end{cases}
I could be wrong though. See this notebook
http://nbviewer.ipython.org/aec7587270c688e701ee

**Comment by bjodah** [comment]
Ooops a typo in the last cell. Update here:
http://nbviewer.ipython.org/654feb2b1d72d9098b3b

so you are right too, even if we get past \begin{cases} we will run into problems when we reach '\\0'.

**Comment by jksuom** [comment]
I think that the failure of ``preview`` is the reason why matplotlib is attempted because of [these lines](https://github.com/sympy/sympy/blob/master/sympy/interactive/printing.py#L124).

**Comment by jksuom** [comment]
It is also true that mathtext in matplotlib only supports a subset of latex. In particular, it can not handle any of the latex environments ``\begin{...}...\end{...}``.

**Comment by bjodah** [comment]
Ah, I see, nice catch.
Here is a workaround:
https://github.com/bjodah/sympy/commit/1d925d1591469e6e5d46a74f95fb32e548749958

But I don't understand why I need four backslashes(!).

**Comment by jksuom** [comment]
Does that work in ipython without notebook?

**Comment by bjodah** [comment]
It works (as in: it produces a sensible image), but it is not correct:

here is the output for Python 2:
```
\documentclass[12pt]{article}
\pagestyle{empty}

\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{euler}

\begin{document}

$\begin{cases} 1 & \text{for}\: x > 0 \\\ 0 & \text{otherwise} \end{cases}$

\end{document}
```
Note the 3 backslashes (the last one escapes the space and includes it so that the 0 is misaligned with respect tot 1). Using Python 3 there are 4 backslashes left.
Which indicates that my escaping before calling `u()` here:
https://github.com/sympy/sympy/blob/master/sympy/printing/preview.py#L194
is not consistent between Py2/Py3

**Comment by jksuom** [comment]
That is what I suspected...
Probably we should add ``backend='dvipng'`` to the ``latex_to_png`` call in ``sympy.interactive.printing``.

**Comment by jksuom** [comment]
[This](http://stackoverflow.com/questions/647769/why-cant-pythons-raw-string-literals-end-with-a-single-backslash) seems to explain why you need four backslashes. (A third one would escape the quote character.)

**Comment by jksuom** [comment]
The null character problem could probably be solved by replacing ``r"\\"`` with ``"\\\\\n"`` (not raw). It should be acceptable to add a newline after the backslashes in latex.

**Comment by jksuom** [comment]
That does not seem to help with the notebook. It does work in SymPy but the resulting latex string will be modified somewhere, probably in ipython, ipykernel or notebook.

**Comment by jksuom** [comment]
It appears that ipython etc. are not to blame. [This line](https://github.com/sympy/sympy/blob/master/sympy/printing/preview.py#L196) will replace a pair of backslashes with a single one spoiling the latex cases string.

I have not been able to find documentation on ``u()``.

**Comment by bjodah** [comment]
Hmm.. I thought I protected pairs of backslashes two lines above by including [r'\\'] in the "preprocessing". `u()` is defined [here](https://github.com/sympy/sympy/blob/master/sympy/core/compatibility.py#L108).

But now I believe I shouldn't have used `u()` in the first place. See #9901

**Comment by derDavidT** [comment]
Please also take note of [this question] (http://stackoverflow.com/questions/30015110/sympy-failed-to-render-matrix/32520117#32520117).

**Comment by bjodah** [comment]
This should have been resolved.

**Comment by derDavidT** [comment]
Hi, i tested it with the current sympy dev version and still get an error; see [this gist] (https://gist.github.com/derDavidT/42a4545ffe09919355d6#file-error_dumpy-py).

However, I'm not sure if it is a sympy or a matplotlib issue...

**Comment by bjodah** [comment]
Did this use to work in an earlier version? If so were you able to bisect what commit introduced the bug?

**Comment by derDavidT** [comment]
Sadly, I just started using matrix variables. I'm sure it is an issue with sympy 0.7.6.1 as well. And looking at the question I posted above it was an issue in May 2015.

**Comment by jksuom** [comment]
This seems to be caused by the ``\begin{smallmatrix}...`` environment that matplotlib does not support.
Will this also happen with ``init_printing(backend='png')``?

**Comment by jksuom** [comment]
Correction: The proper keyword for using the png backend is ``use_latex='png'``, not ``backend='png'`` as I mistakenly suggested.

So I reformulate the question: Does the error occur with this keyword in ``init_printing``? It should bypass matplotlib.

**Comment by derDavidT** [comment]
As described [here] (http://stackoverflow.com/questions/30015110/sympy-failed-to-render-matrix/32520117#32520117), using an external backend does work (since sympy generates valdi Latex code, I guess). My intention was primarily focused on using matplotlib as a backend.

Seeing as matplotlib does not support certain environments, it's up for discussion wether (a) sympy raises a backend error using ``try ... except`` or (b) if it is reasonable to implement a workaround.

**Comment by jksuom** [comment]
Actually, matplotlib does not seem to support any latex environments (with ``\begin...\end{...}``).

**Comment by jksuom** [comment]
I have tried to experiment with various backends using (approximately) the same configuration as in the [dump above](https://gist.github.com/derDavidT/42a4545ffe09919355d6#file-error_dumpy-py):
```
Jupyter QtConsole 4.0.1
Python 2.7.10 |Continuum Analytics, Inc.| (default, May 28 2015, 17:02:03) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
%guiref   -> A brief reference about the graphical user interface.

In [1]: import sympy

In [2]: sympy.init_printing()

In [3]: sympy.Matrix([1, 2])
```

If LaTeX is installed no problems seem to appear. Otherwise,
-  If matplotlib is not installed, there is no error message but no useful output either.

-  If matplotlib is available, the error message (with no other output) is printed.

-  If matplotlib is available and ``init_printing(use_latex=False)`` is applied, there will be no error message and a readable (plain text) output appears.
```
In [4]: sympy.init_printing(use_latex=False)

In [5]: sympy.Matrix([1, 2])
Out[5]: 
⎡1⎤
⎢ ⎥
⎣2⎦
```

These observations suggest the following strategy when ``sympy.init_printing`` is run in a qtconsole or notebook.

If ``use_latex`` is ``None`` (the default), then the existence of LaTeX is tested. If it has been installed, then ``use_latex`` is set to ``True``. Otherwise, it is set to ``False``, and a message is issued telling the user to either install LaTeX or else to set explicitly ``use_latex=False``(to avoid the message).

**Comment by bjodah** [comment]
I think this sounds like a good approach. What do you think @derDavidT ?

**Comment by derDavidT** [comment]
I've got one doubt: When forcing ``use_latex`` to ``False`` if no LaTeX is found, we effectively disable rendering through ``matplotlib`` alltogether...

**Comment by jksuom** [comment]
Yes. It seems to be replaced with the plain text printer. But If one does not want to have ``use_latex=False``, one can choose something else in the ``init_printing()`` call, so nothing is lost. The idea is just to have a default that does not give an error.

**Comment by derDavidT** [comment]
However, we'd have to adapt the sympy documentation, since it specifies that
> If LATEX is not installed, but Matplotlib is installed, it will use the Matplotlib rendering engine. If matplotlib is not installed, it uses the Unicode pretty printer.

**Comment by derDavidT** [comment]
By the way, I've found something rather fishy if LaTex is not installed, but matplotlib is:
Using ``sympy.init_printing()``, we get the expected error if rendering a ``sympy.Matrix``.
Using ``sympy.init_printing(use_latex='matplotlib')``, however, it automatically reverts to unicode printing, if ``matplotlib`` fails.

Following @jksuom initial suggestion: If no LaTeX is found, test for Matplotlib. If it is available set ``use_latex='matplotlib'``.

**Comment by jksuom** [comment]
If ``use_latex`` is ``True`` (the default for notebook and qtconsole) or ``'png'``, SymPy will apply ``_print_latex_png``, which will fall back to ``_matplotlib_wrapper`` if LaTeX is not found. If ``use_latex`` is ``'matplotlib'``, SymPy will use ``_print_latex_matplotlib`` that will also pass the text to ``_matplotlib_wrapper``.

There is a difference, however. The latter function will ``try`` ``_matplotlib_wrapper`` and return ``None`` if it fails. This will probably make ipython revert to plain text.

If the ``try`` statement were in ``_matplotlib_wrapper`` and exceptions from ``latex_to_png`` were caught and replaced with ``None``, it seems that both ``_print_latex_png`` and ``_print_latex_matplotlib`` would behave in the same way.

**Comment by derDavidT** [comment]
I've tried that [here] (https://github.com/derDavidT/sympy), yet it did not solve it. Please have a look, whether I've screwed up somewhere.

**Comment by jksuom** [comment]
@derDavidT Did you see the comments I added to your trial commit?

**Comment by derDavidT** [comment]
@jksuom Have a look at this [gist] (https://gist.github.com/derDavidT/13848c2a1752a4163644).

**Comment by moorepants** [comment]
Using SymPy 0.7.6.1 and IPython 3.2.1 I'm getting this traceback when trying to render matrices with `init_printing()`:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/IPython/core/formatters.pyc in __call__(self, obj)
    328                 pass
    329             else:
--> 330                 return printer(obj)
    331             # Finally look for special method names
    332             method = _safe_get_formatter_method(obj, self.print_method)

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/sympy/interactive/printing.pyc in _print_latex_png(o)
    125                 if latex_mode != 'inline':
    126                     s = latex(o, mode='inline')
--> 127                 return _matplotlib_wrapper(s)
    128 
    129     def _print_latex_matplotlib(o):

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/sympy/interactive/printing.pyc in _matplotlib_wrapper(o)
     88         o = o.replace(r'\operatorname', '')
     89         o = o.replace(r'\overline', r'\bar')
---> 90         return latex_to_png(o)
     91 
     92     def _can_print_latex(o):

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/IPython/lib/latextools.pyc in latex_to_png(s, encode, backend, wrap)
     85     else:
     86         raise ValueError('No such backend {0}'.format(backend))
---> 87     bin_data = f(s, wrap)
     88     if encode and bin_data:
     89         bin_data = encodestring(bin_data)

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/IPython/lib/latextools.pyc in latex_to_png_mpl(s, wrap)
    104     mt = mathtext.MathTextParser('bitmap')
    105     f = BytesIO()
--> 106     mt.to_png(f, s, fontsize=12)
    107     return f.getvalue()
    108 

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/matplotlib/mathtext.pyc in to_png(self, filename, texstr, color, dpi, fontsize)
   3092         image in pixels.
   3093         """
-> 3094         rgba, depth = self.to_rgba(texstr, color=color, dpi=dpi, fontsize=fontsize)
   3095         numrows, numcols, tmp = rgba.shape
   3096         _png.write_png(rgba.tostring(), numcols, numrows, filename)

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/matplotlib/mathtext.pyc in to_rgba(self, texstr, color, dpi, fontsize)
   3057             image in pixels.
   3058         """
-> 3059         x, depth = self.to_mask(texstr, dpi=dpi, fontsize=fontsize)
   3060 
   3061         r, g, b = mcolors.colorConverter.to_rgb(color)

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/matplotlib/mathtext.pyc in to_mask(self, texstr, dpi, fontsize)
   3030         assert(self._output=="bitmap")
   3031         prop = FontProperties(size=fontsize)
-> 3032         ftimage, depth = self.parse(texstr, dpi=dpi, prop=prop)
   3033 
   3034         x = ftimage.as_array()

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/matplotlib/mathtext.pyc in parse(self, s, dpi, prop)
   3003             self.__class__._parser = Parser()
   3004 
-> 3005         box = self._parser.parse(s, font_output, fontsize, dpi)
   3006         font_output.set_canvas_size(box.width, box.height, box.depth)
   3007         result = font_output.get_results(box)

/home/moorepants/miniconda/envs/pydy-examples-py2/lib/python2.7/site-packages/matplotlib/mathtext.pyc in parse(self, s, fonts_object, fontsize, dpi)
   2337                         err.line,
   2338                         " " * (err.column - 1) + "^",
-> 2339                         six.text_type(err)]))
   2340         self._state_stack = None
   2341         self._em_width_cache = {}

ValueError: 
\left ( \left[\begin{smallmatrix}- c v + g m - k x\end{smallmatrix}\right], \quad \left[\begin{smallmatrix}- m \dot{v}\end{smallmatrix}\right]\right )
              ^
Expected "\right" (at char 14), (line:1, col:15)
```

We really need to take care of this issue. It is quite annoying and there are a bunch of issues in the tracker about it.

**Comment by jksuom** [comment]
This issue only appears under some special conditions. I tried with IPython 3.2.1:
```
IPython console for SymPy 0.7.6.1 (Python 2.7.10-64-bit) (ground types: python)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at http://www.sympy.org

In [1]: Matrix([1, 2])
Out[1]: 
⎡1⎤
⎢ ⎥
⎣2⎦
```
There must be some invisible difference in our systems.

**Comment by bjodah** [comment]
@moorepants agreed, the problem is that matplotlib does not have full support for LaTeX.
See @derDavidT's gist and the following discussion above.

**Comment by moorepants** [comment]
My traceback was from the IPython Notebook. I seemingly have LaTeX installed, so I wouldn't expect this issue. It hasn't come up for me in the past if I had LaTeX installed.

**Comment by moorepants** [comment]
I know that matplotlib doesn't have support for all LaTeX, but I think the actual problem is that we compute the matplotlib rendering when it isn't needed. The matplotlib calls should only run if:

1. You explicitly ask for matplotlib lib.
2. You are in an environment (qtconsole, notebook) and you don't have mathjax or latex available.

Otherwise these calls to matplotlib should never happen. Why are we creating a png from matplotlib if it is never used?

**Comment by jksuom** [comment]
I installed ipython-notebook-3.2.1 but even that did not produce the error. Maybe MathJax was used:
```
Using MathJax from CDN: https://cdn.mathjax.org/mathjax/latest/MathJax.js
```
@moorepants Are you perhaps running offline?

**Comment by moorepants** [comment]
I just filed this issue: https://github.com/matplotlib/matplotlib/issues/5162

I couldn't find any similar issues on their tracker.

@jksuom I am not running offline. I'm pretty sure that the png is rendered in the background regardless if MathJax is present or not.

**Comment by moorepants** [comment]
This shows that both the png is rendered by matplotlib and that it ultimately gets rendered by MathJax:
![selection_004](https://cloud.githubusercontent.com/assets/276007/10221347/ad25c340-6802-11e5-9224-b13936a513dc.png)

**Comment by jksuom** [comment]
> preview failed

Is that because latex was not found?

**Comment by moorepants** [comment]
That's because dvipng was not found. Here is the notebook with debug mode. Notice how both the png and mathjax renderers are loaded simultaneously. It seems like if MathJax is available then we shouldn't load the png renderer.
![selection_005](https://cloud.githubusercontent.com/assets/276007/10221538/dac5cec0-6803-11e5-99fb-0f2af007319a.png)

**Comment by moorepants** [comment]
This could be an IPython limitation. IPython is structured such that you can't easily and reliably detect which environment code was executed in. For things to work in qtconsole we always need png output but for the notebook we need mathjax or png, but there isn't any way to know if the code was executed in the qtconsole or the notebook...I think.

**Comment by moorepants** [comment]
One potential solution here would be to switch from the amsmath matrix environments to array, see:

http://stackoverflow.com/questions/26329177/how-to-display-a-matrix-in-the-matplotlib-annotations

But we need to make sure MathJax renders array correctly too.

**Comment by moorepants** [comment]
Scrap that. mathtext does not support any LaTeX environments (`\begin{}\end{}`). See: https://github.com/matplotlib/matplotlib/issues/5162#issuecomment-144728549

Here are some options:
- Patch mathtext to support latex environments
- Raise a sympy error if the latex command includes latex environments and matplotlib is the only available rendered (still need to deal with simultaneously loading png and mathjx in notebook).
- Detect if the code was executed in the terminal, qtconsole, or notebook and for the later whether mathjax is installed so that we can choose the single best formatter (e.g. in the notebook choose mathjax by default).
- Detect if MathJax is available for all ipython envs and default to using it. (this likely requires us to know the Ipython env too).

**Comment by jksuom** [comment]
I am still unable to produce the error even after making latex unavailable. @moorepants What happens with matrix printing if you run ``init_printing(use_latex='matplotlib')``?

**Comment by moorepants** [comment]
Fixed in #9959.
